c_glib/arrow-glib/decimal.cpp

```diff
@@ -0,0 +1,185 @@
+  PROP_DECIMAL128
+};
+
+G_DEFINE_TYPE_WITH_PRIVATE(GArrowDecimal128,
+                           garrow_decimal128,
+                           G_TYPE_OBJECT)
+
+#define GARROW_DECIMAL128_GET_PRIVATE(obj)                 \
+  (G_TYPE_INSTANCE_GET_PRIVATE((obj),                   \
+                               GARROW_TYPE_DECIMAL128,     \
+                               GArrowDecimal128Private))
+
+static void
+garrow_decimal128_finalize(GObject *object)
+{
+  auto priv = GARROW_DECIMAL128_GET_PRIVATE(object);
+
+  priv->decimal128 = nullptr;
+
+  G_OBJECT_CLASS(garrow_decimal128_parent_class)->finalize(object);
+}
+
+static void
+garrow_decimal128_set_property(GObject *object,
+                        guint prop_id,
+                        const GValue *value,
+                        GParamSpec *pspec)
+{
+  auto priv = GARROW_DECIMAL128_GET_PRIVATE(object);
+
+  switch (prop_id) {
+  case PROP_DECIMAL128:
+    priv->decimal128 =
+      *static_cast<std::shared_ptr<arrow::Decimal128> *>(g_value_get_pointer(value));
+    break;
+  default:
+    G_OBJECT_WARN_INVALID_PROPERTY_ID(object, prop_id, pspec);
+    break;
+  }
+}
+
+static void
+garrow_decimal128_init(GArrowDecimal128 *object)
+{
+}
+
+static void
+garrow_decimal128_class_init(GArrowDecimal128Class *klass)
+{
+  GParamSpec *spec;
+
+  auto gobject_class = G_OBJECT_CLASS(klass);
+
+  gobject_class->finalize     = garrow_decimal128_finalize;
+  gobject_class->set_property = garrow_decimal128_set_property;
+
+  spec = g_param_spec_pointer("decimal128",
+                              "Decimal128",
+                              "The raw std::shared<arrow::Decimal128> *",
+                              static_cast<GParamFlags>(G_PARAM_WRITABLE |
+                                                       G_PARAM_CONSTRUCT_ONLY));
+  g_object_class_install_property(gobject_class, PROP_DECIMAL128, spec);
+}
+
+/**
+ * garrow_decimal128_new_string:
+ * @data: The data of the decimal.
+ *
+ * Returns: A newly created #GArrowDecimal128.
+ *
+ * Since: 1.0.0
+ */
+GArrowDecimal128 *
+garrow_decimal128_new_string(const gchar *data)
+{
+  auto arrow_decimal = std::make_shared<arrow::Decimal128>(data);
+  return garrow_decimal128_new_raw(&arrow_decimal);
+}
+
+/**
+ * garrow_decimal128_new_integer:
+ * @data: The data of the decimal.
+ *
+ * Returns: A newly created #GArrowDecimal128.
+ *
+ * Since: 1.0.0
+ */
+GArrowDecimal128 *
+garrow_decimal128_new_integer(const gint64 data)
+{
+  auto arrow_decimal = std::make_shared<arrow::Decimal128>(data);
+  return garrow_decimal128_new_raw(&arrow_decimal);
+}
+
+/**
+ * garrow_decimal128_to_string:
+ * @decimal: A #GArrowDecimal128.
+ * @scale: The scale of the decimal.
+ *
+ * Returns: The string representation of the decimal.
+ *
+ * Since: 1.0.0
+ */
+const gchar *
+garrow_decimal128_to_string(GArrowDecimal128 *decimal, gint32 scale)
+{
+  auto arrow_decimal = garrow_decimal128_get_raw(decimal);
+  return arrow_decimal->ToString(scale).c_str();
```

Comment:
This is not safe because `ToString()` returns a temporary `std::string`. It's unavailable outside of this scope.

We need to return duplicated `gchar *`:

```diff
diff --git a/c_glib/arrow-glib/decimal.cpp b/c_glib/arrow-glib/decimal.cpp
index f5934cc9..37a88e54 100644
--- a/c_glib/arrow-glib/decimal.cpp
+++ b/c_glib/arrow-glib/decimal.cpp
@@ -142,13 +142,16 @@ garrow_decimal128_new_integer(const gint64 data)
  *
  * Returns: The string representation of the decimal.
  *
+ *   It should be freed with g_free() when no longer needed.
+ *
  * Since: 1.0.0
  */
-const gchar *
+gchar *
 garrow_decimal128_to_string(GArrowDecimal128 *decimal, gint32 scale)
 {
   auto arrow_decimal = garrow_decimal128_get_raw(decimal);
-  return arrow_decimal->ToString(scale).c_str();
+  auto string = arrow_decimal->ToString(scale);
+  return g_strndup(string.data(), string.size());
 }
 
 /**
```

```diff
diff --git a/c_glib/arrow-glib/decimal.h b/c_glib/arrow-glib/decimal.h
index 4c1ef545..8245abc5 100644
--- a/c_glib/arrow-glib/decimal.h
+++ b/c_glib/arrow-glib/decimal.h
@@ -37,8 +37,8 @@ struct _GArrowDecimal128Class
 
 GArrowDecimal128 *garrow_decimal128_new_string(const gchar *data);
 GArrowDecimal128 *garrow_decimal128_new_integer(const gint64 data);
-const gchar *garrow_decimal128_to_string(GArrowDecimal128 *decimal,
-                                         gint32 scale);
+gchar *garrow_decimal128_to_string(GArrowDecimal128 *decimal,
+                                   gint32 scale);
 const gchar *garrow_decimal128_to_integer_string(GArrowDecimal128 *decimal);
 
 G_END_DECLS
```